src/iomap.cpp

```diff
@@ -31,209 +31,118 @@
-				if (!ok) {
-					setLastErrorString("Invalid spawn tag.");
-					return false;
-				}
-
-				map.spawnfile = fileName.parent_path() / spawnFile;
+				spawns = OTB::readString(first, last);
 				break;
 			}
 
 			case OTBM_ATTR_EXT_HOUSE_FILE: {
-				auto [houseFile, ok] = propStream.readString();
-				if (!ok) {
-					setLastErrorString("Invalid house tag.");
-					return false;
-				}
-
-				map.housefile = fileName.parent_path() / houseFile;
+				houses = OTB::readString(first, last);
 				break;
 			}
 
 			default:
-				setLastErrorString("Unknown header node.");
-				return false;
+				throw std::invalid_argument(fmt::format("Unknown map attribute {:d}\n", attr));
 		}
 	}
-	return true;
+
+	return std::make_pair(spawns, houses);
 }
 
-bool IOMap::parseTileArea(OTB::Loader& loader, const OTB::Node& tileAreaNode, Map& map)
+auto read_coords(auto& first, auto const last)
 {
-	PropStream propStream;
-	if (!loader.getProps(tileAreaNode, propStream)) {
-		setLastErrorString("Invalid map node.");
-		return false;
+	auto x = OTB::read<uint16_t>(first, last);
+	auto y = OTB::read<uint16_t>(first, last);
+	auto z = OTB::read<uint8_t>(first, last);
+	return std::make_tuple(x, y, z);
+}
+
+Tile* createTile(Item*& ground, Item* item, uint16_t x, uint16_t y, uint8_t z)
+{
+	if (!ground) {
+		return new StaticTile(x, y, z);
 	}
 
-	OTBM_Destination_coords area_coord;
-	if (!propStream.read(area_coord)) {
-		setLastErrorString("Invalid map node.");
-		return false;
+	Tile* tile;
+	if ((item && item->isBlocking()) || ground->isBlocking()) {
+		tile = new StaticTile(x, y, z);
+	} else {
+		tile = new DynamicTile(x, y, z);
 	}
 
-	uint16_t base_x = area_coord.x;
-	uint16_t base_y = area_coord.y;
-	uint16_t z = area_coord.z;
+	tile->internalAddThing(ground);
+	ground->startDecaying();
+	ground = nullptr;
+	return tile;
+}
 
-	for (auto& tileNode : tileAreaNode.children) {
-		if (tileNode.type != OTBM_TILE && tileNode.type != OTBM_HOUSETILE) {
-			setLastErrorString("Unknown tile node.");
-			return false;
-		}
+void parseTileArea(const OTB::Node& node, Map& map)
+{
+	auto node_begin = node.propsBegin;
+	auto [base_x, base_y, z] = read_coords(node_begin, node.propsEnd);
 
-		if (!loader.getProps(tileNode, propStream)) {
-			setLastErrorString("Could not read node data.");
-			return false;
+	for (auto& tileNode : node.children) {
+		if (tileNode.type != OTBM_TILE && tileNode.type != OTBM_HOUSETILE) [[unlikely]] {
+			throw std::invalid_argument(fmt::format("Unknown tile node: {:d}.", static_cast<uint16_t>(node.type)));
 		}
 
-		OTBM_Tile_coords tile_coord;
-		if (!propStream.read(tile_coord)) {
-			setLastErrorString("Could not read tile position.");
-			return false;
-		}
+		auto first = tileNode.propsBegin, last = tileNode.propsEnd;
 
-		uint16_t x = base_x + tile_coord.x;
-		uint16_t y = base_y + tile_coord.y;
+		uint16_t x = base_x + OTB::read<uint8_t>(first, last);
+		uint16_t y = base_y + OTB::read<uint8_t>(first, last);
```

Comment:
static_cast + auto could be an alternative?

Reply:
Cast not needed, it is promoted automatically